### ISSUE TITLE
[Feature Request] Feature and Trajectory Interface

### ISSUE THREAD

**Issue by WolfByttner** [patrickfletcher/clODE#37]
**New interfaces for the base classes**

clODE is a parallel dynamical systems library that specialises in large-scale simulations. It fills a critical gap in the Python ecosystem.

The interface should ideally follow scipy.solve_ivp - at the very least, the right-hand side equation should.

Below is the current state of the proposed API.

```python
import numpy as np
import matplotlib.pyplot as plt
import clode
from typing import List

# The FitzHugh-Nagumo model

def fitzhugh_nagumo(time: float,
                    variables: List[float],
                    a: float,
                    b: float,
                    current: float,
                    epsilon: float) -> List[float]:
    v: float = variables[0]
    w: float = variables[1]
    dv = v - v ** 3 / 3 - w + current
    dw = epsilon * (v + a - b * w)
    return [dv, dw]

a = [0.7, 0.8, 0.9, 1.0]

simulator = clode.FeatureSimulator(
    rhs=fitzhugh_nagumo,
    variables={"v": np.arange(-2, 2, 0.2), "w": -1.0},
    parameters={"a": a, "b": 0.8, "epsilon": 0.01, "current": 1.0},
)

simulator.transient(t_span=(0.0, 100.0))
observer_output = simulator.features()

voltages = observer_output.get_var("v")

amplitude = observer_output.get_var_max("v") - observer_output.get_var_min("v")
period_min = observer_output.get_var_min("period")

# Object classes, silent=0, excitable=1, spiking=2
classes = np.zeros(len(voltages))

classes[amplitude > 20] = 1  # excitable
classes[(amplitude > 20) & (period_min > 5)] = 2  # spiking

# Plotting
plt.figure(figsize=(8, 6))
plt.scatter(a, voltages, c=classes, cmap="viridis")

plt.show()
```

**Comment by patrickfletcher** [comment]
@WolfByttner this is looking good. Perhaps the variables/parameters inputs to the Simulator constructors should be only the name+default value dicts?  I'm guessing you're thinking to place the "initialze" call inside the constructor. We should decide on a convention here - what makes most sense in terms of where to pass tspan, X0, and the pars array? A key point is that setting X0+pars sets the number of simulations, and therefore specifies the OpenCL buffer sizes needed. Initialize actually allocates those arrays. Options include:
- constructor (not my favorite, as those are "experiment-specific"... this may still be a good place, at least as an overload)
- separate "initialize" call (for X0 and pars) - useful if you want to restart, say with a different number of simulations. This can still be available, even if we use another convenience overload...
- the simulation calls (transient, features, trajectory) - lazy initialization, done just before the simulation is actually run. 

I think it is probably most intuitive to support passing in tspan to the simulation functions. I'm happy with keeping a separate "initialize" call for setting X0 and pars - setting those is a purposeful action where you declare the size of your ensemble.

Thoughts?

**Comment by patrickfletcher** [comment]
The two clode-specific features are Aux and Wiener variables, neither of which are supported by Scipy.  Here are some specific thoughts about aux/Wiener that should help clarify how to treat these:

- Aux and Wiener variables appear as elements of a write-only (aux) or read-only (Wiener) array. 
- Aux variables are defined as expressions that may reference any local variable in the get_rhs function. However, they must never be referenced by any expression that ultimately impacts the derivatives. They are read-outs only. That is, the aux array is write-only, and can be computed after all the derivatives.
- Wiener variables are read-only inputs (note the "const" in the get_rhs signature). They may be assigned to a local variable for readability, and they are used in derivative computation. Scipy doesn't solve problems of this type (stochastic differential equations - SDEs) but there is a package called sdeint. Let's not worry about supporting that here.

Given the above, we could have the user-defined get_rhs function adopt a Scipy-compatible signature. The parameters could be listed as a list of required arguments (*args) or as a vector like vars. With the former version, parameter default values could be listed in the argument list.  Wiener variables (if specified) could appear as the next input argument(s) with default value(s), so that their values would be constant for the purposes of Scipy.  For clode, the user would specify their names in a list, so the converter knows to treat them as such.  

For aux variables, we could treat them with the conditional return I suggested before, unless you see a more elegant/readable solution. Essentially, we need Scipy *not* to see them, but to expose their expressions in the CL RHS file. Note that it is also useful to be able to compute them post-hoc, once a trajectory (t, y) has been computed via scipy.  That is, just pass the trajectory back through the RHS function to compute the resulting derivatives and aux vectors.

In summary, how does something like the following look?

```python
def my_rhs(t, y, pars, w1=0.0, return_aux = False):
    ... #compute derivatives
    derivatives = [xprime, yprime]
    if return_aux:
        ... #compute aux expressions
        aux = [expr1, expr2]
        return derivatives, aux
    else:
        return derivatives

pars = {"p1":1.0, "p2":1.7} #parameter metadata
vars = {"x":0.0, "y":1.2}   #variable metadata
aux = ["aux1", "aux2"]
wvars = ["w1"]

simulator = SimulatorBase(my_rhs, pars, vars, aux, wvars, solver_pars)
... #create parameter + initial condition matrices, P, X0
simulator.initialize(P, X0) #allocates GPU memory
simulator.transient(t_span=(0.0, 100.0))
xf = simulator.get_final_state()
```

With this API, multiple integrator instances could be created from the same metadata, but would have independent state. It would be up to the user to manage them.

I favor keeping the "initialize" routine, or possibly reworking it to use "setProblemData" instead (a C++ level change), as this is a key step: it tells the system how much device memory needs to be allocated. Specifically, the number of simulations is coded in the number of rows of the P and X0 matrices.  One may wish to keep the same simulator object and change this number on the fly, as opposed to creating a new simulator.

**Comment by patrickfletcher** [comment]
closed by #32
